Feat: Support repo specific claude-code session

tninja opened the pull request:
This PR is about to let claude-code.el launch repo specific claude-code session, so that people can work on different repo with claude-code in emacs.

@stevemolitor PTAL Thanks

stevemolitor commented:
This was on my todo list so this is really great, thanks @tninja!

I’m on vacation without a computer so I won’t be able to take a closer look for a couple weeks. But overall it looks good on my phone.

I made some changes on main anticipating this feature that might cause some minor conflicts so you might want to rebase or merge to resolve. It should reduce the size of this MR.

Instead of using ‘vc-git’ I wonder if it’d be better to do something like ‘project-root (project-current t))’ so it will work with other version control systems also.

Thanks!

tninja commented:
It is great that you like it. Let me know when it is a good time to merge or rebase main branch.

I'll check the project-root function, and see how to replace vc-git.

tninja commented:
comment addressed

stevemolitor commented:
@tninja I pushed some changes:

- rebased to pull in latest `claude-code.el` changes
- fixed bug where it prompted for the project directory when it didn't need to. This was introduced by my suggestion to use `(project-current t)` - where `t` means to prompt if no project. I removed the `t`. My bad sorry.
- some minor cleanup and refactoring.
- Removed `claude-code-current-directory`. I don't think it's necessary know as `claude-code` will just do the right thing with your changes. You can still force it to prompt for a project directory with a prefix arg.
- Updated README and CHANGELOG.

I'm going to test it out for a day or to - let me know if you see any issues.

Thanks so much for adding this feature! It's very useful and something I had wanted to add, but you saved me a bunch of time. Thanks!

stevemolitor commented:
Two usability issues I noticed while using this branch:

- The "Claude is not running" message that is shown when you run, e.g., `claude-kill`, `claude-toggle`, etc, can be a little misleading when there is a Claude process running but for a different project. I fixed that in e392bf419e9fb00449ab13747faabf751c225f9f. Now it says "Claude is not running in directory /foo".
- Sometimes, while working on a project, I like to work in a sibling directory that has a dependency or project notes. I'd still like to send commands to Claude, toggle the Claude buffer, etc. However, now I get the "Claude is not running in ..." message. I'm thinking of adding a facility to let the user pick which Claude process to use if there are other processes running, and if there is just one Claude process running, maybe defaulting to that. I'll work on this over the next few days.

stevemolitor commented:
OK I've updated this PR with a workflow that works for me. I sometimes like to work in sibling directories, where a sibling directory is the source for a dependency of the main project I'm working on, or I mess around in the *scratch* buffer and want to interact with the Claude process for the project I'm working on. I've added changes to let the user pick which Claude instance they want to work with if they invoke a `claude-code.el` command outside of the directory for a running Claude process, and remember their selection so as not to be too annoying. 

Here is the updated [READMD](https://github.com/stevemolitor/claude-code.el/blob/57788b11cde64bbf9e88043911718da5b062ab7a/README.md#multiple-claude-instances):

### Multiple Claude Instances

`claude-code.el` supports running multiple Claude instances across different projects and directories. Each Claude instance is associated with a specific directory (project root, file directory, or current directory).

#### Instance Management

- When you start Claude with `claude-code`, it creates an instance for the current directory
- Buffer names follow the format `*claude:/path/to/directory*` to clearly identify each instance
- If you're in a directory without a Claude instance but have instances running in other directories, you'll be prompted to select one
- Your selection is remembered for that directory, so you won't be prompted again
- To start a new instance instead of selecting an existing one, cancel the prompt with `C-g`

This allows you to have separate Claude conversations for different projects while easily switching between them or sharing a Claude instance across related directories.

tninja commented:
Thank you. This is a huge amount of work!

On Fri, Jun 6, 2025 at 3:22 PM Steve Molitor ***@***.***>
wrote:

> *stevemolitor* left a comment (stevemolitor/claude-code.el#12)
> <https://github.com/stevemolitor/claude-code.el/pull/12#issuecomment-2950943985>
>
> OK I've updated this PR with a workflow that works for me. I sometimes
> like to work in sibling directories, where a sibling directory is the
> source for a dependency of the main project I'm working on, or I mess
> around in the *scratch* buffer and want to interact with the Claude
> process for the project I'm working on. I've added changes to let the user
> pick which Claude instance they want to work with if they invoke a
> claude-code.el command outside of the directory for a running Claude
> process, and remember their selection so as not to be too annoying.
>
> I will test this out for a day or two before merging. Feedback and/or bug
> reports welcome.
>
> Thanks again @tninja <https://github.com/tninja>!
>
> Here is the updated README
> <https://github.com/stevemolitor/claude-code.el/blob/57788b11cde64bbf9e88043911718da5b062ab7a/README.md#multiple-claude-instances>
> :
> Multiple Claude Instances
>
> claude-code.el supports running multiple Claude instances across
> different projects and directories. Each Claude instance is associated with
> a specific directory (project root, file directory, or current directory).
> Instance Management
>
>    - When you start Claude with claude-code, it creates an instance for
>    the current directory
>    - Buffer names follow the format *claude:/path/to/directory* to
>    clearly identify each instance
>    - If you're in a directory without a Claude instance but have
>    instances running in other directories, you'll be prompted to select one
>    - Your selection is remembered for that directory, so you won't be
>    prompted again
>    - To start a new instance instead of selecting an existing one, cancel
>    the prompt with C-g
>
> This allows you to have separate Claude conversations for different
> projects while easily switching between them or sharing a Claude instance
> across related directories.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stevemolitor/claude-code.el/pull/12#issuecomment-2950943985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOPAILKP5LBFQZGQMCDQD3CIIDHAVCNFSM6AAAAAB537APP2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNJQHE2DGOJYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
